File: packages/expo-brownfield/android/src/main/java/expo/modules/brownfield/ExpoBrownfieldStateModule.kt
package expo.modules.brownfield

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

const val KEY_RECREATED_EVENT_NAME = "onKeyRecreated"

class ExpoBrownfieldStateModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoBrownfieldStateModule")

    Events(KEY_RECREATED_EVENT_NAME)

    Class(SharedState::class) {
      Constructor { key: String -> SharedState(key) }

      Function("get") { state: SharedState ->
        return@Function state.get()
      }

      Function("set") { state: SharedState, value: Any? -> state.set(value) }
    }

    OnCreate { BrownfieldState.setExpoModule(this@ExpoBrownfieldStateModule) }

    OnDestroy { BrownfieldState.setExpoModule(null) }

    Function("getSharedState") { key: String ->
      return@Function BrownfieldState.getOrCreate(key)
    }

    Function("deleteSharedState") { key: String -> BrownfieldState.delete(key) }
  }

  fun notifyKeyRecreated(key: String) {
    sendEvent(KEY_RECREATED_EVENT_NAME, mapOf("key" to key))
  }
}

File: packages/expo-brownfield/ios/ExpoBrownfieldStateModule.swift
import ExpoModulesCore

let KEY_RECREATED_EVENT_NAME = "onKeyRecreated"

// MARK: - ExpoBrownfieldStateModule

public class ExpoBrownfieldStateModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoBrownfieldStateModule")

    Events(KEY_RECREATED_EVENT_NAME)

    Class(SharedState.self) {
      Constructor { (key: String) in
        return SharedState(key)
      }

      Function("get") { (state: SharedState) -> Any? in
        return state.get()
      }
      
      Function("set") { (state: SharedState, value: JavaScriptValue?) in
        state.set(value?.getRaw())
      }
    }

    OnCreate {
      BrownfieldStateInternal.shared.setExpoModule(self)
    }

    OnDestroy {
      BrownfieldStateInternal.shared.setExpoModule(nil)
    }

    Function("getSharedState") { (key: String) -> SharedState in
      return BrownfieldStateInternal.shared.getOrCreate(key)
    }

    Function("deleteSharedState") { (key: String) in
      BrownfieldStateInternal.shared.delete(key)
    }
  }

  public func notifyKeyRecreated(_ key: String) {
    sendEvent(KEY_RECREATED_EVENT_NAME, ["key": key])
  }
}
